Compose UI

Lvntt opened the pull request:
Added compose UI & MVVM

Lvntt committed:
fixed styles

removed margins and dimens from styles

Lvntt committed:
changed colors' names

replaced colors' names with ones from figma layout

Lvntt committed:
made <shape> a root element

Lvntt committed:
renamed string sum to addition

Lvntt committed:
added digit and operations enums

Lvntt committed:
refactored code, fixed some bugs, added livedata

changed binding init, added string constants, moved companion objects to the bottom, added extra functions to have better readability, added better exception handling, fixed incorrect behaviour with irrational numbers, added livedata

Lvntt committed:
added compose ui

added compose dependencies, compose-related files for styling etc., calculator layout, changed enum to a more convenient one, added buttons source

Lvntt committed:
added mvvm and text style for error

Lvntt committed:
added ui state and changed error handling

Chantreck commented on app/src/main/res/drawable/button_shape.xml:
```diff
@@ -1,8 +1,5 @@
-    </item>
-</selector>
\ No newline at end of file
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
```
Если ты на компоуз перешёл, зачем этот файл?)

Chantreck commented on app/src/main/java/com/example/calculator/MainActivity.kt:
```diff
@@ -1,117 +1,258 @@
+        EraseIcon(viewModel = viewModel)
+        CalculatorDivider()
+        AllCalculatorButtons(
+            viewModel = viewModel,
```
Не нужно передавать `viewModel` в компонент. Либо у компонента есть своя `ViewModel` (обычно, если это либо компонент экрана, либо компонент со сложной логикой), либо в компонент передаётся только лямбда для `onClick`. Можно передать либо непосредственно в виде лямбды, либо через `Modifier.clickable()` с `when` внутри (я больше за первый вариант в данном случае). Делать так нужно, потому что кнопка не должна знать, что существует какая-то вью-модель, которая будет обрабатывать нажатие кнопки. Кнопка просто говорит: вот я могу нажаться, при нажатии нужно сделать какое-то действие, всё остальное её мало волнует. К тому же одни и те же кнопки могут быть на разных экранах, а значит у них будут разные вью-модели

Chantreck commented on app/src/main/java/com/example/calculator/MainActivity.kt:
```diff
@@ -1,117 +1,258 @@
-            binding.error.visibility = View.VISIBLE
+@Composable
+fun ExpressionText(
+    viewModel: CalculatorViewModel,
```
Здесь можно обойтись без передачи всей вью-модели, достаточно просто передать UiState и выражение

Chantreck commented on app/src/main/java/com/example/calculator/MainActivity.kt:
```diff
@@ -1,117 +1,258 @@
-            updateUi()
+@Composable
+fun EraseIcon(
+    viewModel: CalculatorViewModel,
```
Здесь заменяем вью-модель на `onIconClick: () -> Unit`

Chantreck commented on app/src/main/java/com/example/calculator/MainActivity.kt:
```diff
@@ -1,117 +1,258 @@
+    Row(
+        modifier = modifier.fillMaxWidth(),
+    ) {
+        if (buttonsRow.size == 3) {
```
В идеале здесь тоже проходиться циклом по кнопкам. Условие на размер ряда нужно только для 1 или 1-2 кнопок, остальные кнопки никак от него не зависят, надо их вынести из условия

Chantreck commented on app/src/main/java/com/example/calculator/MainActivity.kt:
```diff
@@ -1,117 +1,258 @@
+
+@Composable
+fun CalculatorButton(
+    viewModel: CalculatorViewModel,
```
Здесь совершенно точно не нужна вью-модель, достаточно просто заменить на `onClick: (CalculatorButton) -> Unit`

Chantreck commented on app/src/main/java/com/example/calculator/MainActivity.kt:
```diff
@@ -1,117 +1,258 @@
+        CalculatorDivider()
+        AllCalculatorButtons(
+            viewModel = viewModel,
+            buttonsGrid = ButtonsSource.buttons
```
Если используешь MVVM, то в идеале получить кнопки от `ViewModel`, которая будет брать их из модели (например, как у тебя из `ButtonSource`)

Chantreck commented on app/src/main/java/com/example/calculator/model/CalculatorModel.kt:
```diff
@@ -0,0 +1,192 @@
+
+    private var canResetExpression = false
+    private val operationSigns = listOf(
+        CalculatorButton.ADDITION.value[0],
```
Чтобы не делать так, у `value` можно изначально задать тип `Char`

Chantreck commented on app/src/main/java/com/example/calculator/presentation/CalculatorUiState.kt:
```diff
@@ -0,0 +1,10 @@
+
+    object Initial : CalculatorUiState
+
+    object Input : CalculatorUiState
```
В этих объектах можно было бы инкапсулировать сразу и значение выражения, поскольку у тебя `UiState` всегда будет обновляться вместе с `currentExpression`, поэтому логично иметь их вместе. Для этого можно `object`, например, заменить на `data class`. Для ошибки можно даже задать дефолтное значение сообщения, раз оно у тебя как правило не меняется

Chantreck commented on app/src/main/java/com/example/calculator/presentation/CalculatorViewModel.kt:
```diff
@@ -0,0 +1,71 @@
+            CalculatorButton.MULTIPLICATION,
+            CalculatorButton.DIVISION
+            -> {
+                _currentExpression.value = calculatorModel.handleOperationClick(button.value)
```
А зачем передавать в модель строку? Передавай тот же CalculatorButton, опять-таки обезопасишь handleOperationClick, поскольку сразу будет сильно ограничены возможные входные данные, а если надо получить сам знак, уже внутри модели можешь его взять.

Chantreck commented on app/src/main/java/com/example/calculator/model/CalculatorModel.kt:
```diff
@@ -0,0 +1,192 @@
+    fun handleOperationClick(operation: String): String {
+        if (currentExpression.isEmpty()) return currentExpression
+        if (operationSignCanBeReplaced()) {
+            val chars = currentExpression.toCharArray()
+            chars[chars.lastIndex] = operation[0]
+            currentExpression = chars.joinToString(EMPTY_STRING)
+            canResetExpression = false
```
Всё же кажется, что если хранить операнды по отдельности, то не нужно будет придумывать такие обходы в виде `CharArray` и прочего, выглядит очень экстравагантно)

Lvntt replied:
Не спорю, делать через регулярку было не лучшей идеей, но почему-то захотелось в качестве эксперимента.

Chantreck reviewed (changes requested):
На этот раз из-за перехода на компоуз пришлось начинать практически с нуля, поэтому логику и стиль кода не особо смотрел, оставим на следующий раз

Lvntt committed:
issues fix

changed ui state interface, deleted extra files, handled viewmodel ifno differently, added an initial text style, formatted code in model

Lvntt committed:
issues fix #2

added dynamic theming

Lvntt committed:
issues fix #3

fixed font on the buttons

Lvntt committed:
issues fix #4

added app bars color and vertical scroll

Chantreck commented on app/src/main/java/com/example/calculator/MainActivity.kt:
```diff
@@ -1,117 +1,256 @@
+import com.example.calculator.presentation.CalculatorUiState
+import com.example.calculator.presentation.CalculatorViewModel
+import com.example.calculator.ui.theme.*
+class MainActivity : ComponentActivity() {
```
Пробельчика не хватает перед объявлением класса

Chantreck commented on app/src/main/java/com/example/calculator/MainActivity.kt:
```diff
@@ -1,117 +1,256 @@
+        EraseIcon(viewModel::handleEraseIconClick)
+        CalculatorDivider()
+        AllCalculatorButtons(
+            buttonsGrid = viewModel.getButtons(),
```
Здесь наверное тоже стоит по ссылке передавать, раз ты в остальных местах так делаешь

Chantreck commented on app/src/main/java/com/example/calculator/MainActivity.kt:
```diff
@@ -1,117 +1,256 @@
-        val currentExpression = calculator.getCurrentExpression()
+@Composable
+fun AppTitleText(modifier: Modifier = Modifier) {
+    Row(modifier = modifier) {
```
По-моему `Row` лишний, можно было бы без него, чтобы лишних вложенностей не было

Chantreck commented on app/src/main/java/com/example/calculator/MainActivity.kt:
```diff
@@ -1,117 +1,256 @@
+            bottom = ExpressionTextBottomPadding
+        )
+    ) {
+        when (calculatorUiState) {
```
Красивое решение, я прям кайфанул)

Chantreck commented on app/src/main/java/com/example/calculator/presentation/CalculatorViewModel.kt:
```diff
@@ -0,0 +1,77 @@
+
+    private val calculatorModel = CalculatorModel()
+
+    private fun setError() {
```
Обычно сначала размещают все публичные функции, а затем уже защищенные и приватные (обычно в порядке вызова из публичных)

Chantreck commented on app/src/main/java/com/example/calculator/model/CalculatorModel.kt:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    companion object {
+        private const val TAG = "CALCULATOR"
```
Чтобы не писать кучу раз private, можно завести отдельно `private companion object`, тогда все его члены автоматически будут приватными. И да, у класса может быть несколько companion object, но ERROR_MESSAGE возможно есть смысл просто вынести в отдельный объект, так как используется несколькими классами

Lvntt committed:
final refactoring

moved private methods to the bottom, made companion object private, added constants file with error message in it, removed redundant row
